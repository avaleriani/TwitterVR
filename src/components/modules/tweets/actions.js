import moment from 'moment';
import { v4 as uuid } from 'uuid';

const scope = '@@tweets';

export const CREATE_TWEET = `${scope}/CREATE_TWEET`;

export const createTweet = payload => ({
  type: CREATE_TWEET,
  payload: {
    id: uuid(), // unique id of tweet
    createdAt: moment(), // datetime object of tweet's creation
    userId: null, // user's id
    text: '', // content of the tweet
    replyToId: null, // if the tweet is a reply, id to the original tweet
    ...payload,
  },
});
