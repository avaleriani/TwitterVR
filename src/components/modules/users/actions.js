import { v4 as uuid } from 'uuid';
import * as fromUsers from './reducer';

const scope = '@@users';

export const RECIEVE_AUTH = `${scope}/RECIEVE_AUTH`;
export const SIGNIN = `${scope}/SIGNIN`;
export const SIGNUP = `${scope}/SIGNUP`;
export const LOGIN = `${scope}/LOGIN`;
export const LOGOUT = `${scope}/LOGOUT`;

export const recieveAuth = () => ({
  type: RECIEVE_AUTH,
});

export const signin = id => ({
  type: SIGNIN,
  payload: {
    id,
  },
});

export const signup = username => ({
  type: SIGNUP,
  payload: {
    id: uuid(),
    username,
  },
});

export const login = username => (dispatch, getState) => {
  const { users } = getState();

  dispatch({ type: LOGIN, username });

  const userId = fromUsers.getUserIdByUsername(users, username);

  if (userId) {
    return dispatch(signin(userId));
  }

  return dispatch(signup(username));
};

export const logout = () => ({
  type: LOGOUT,
});
