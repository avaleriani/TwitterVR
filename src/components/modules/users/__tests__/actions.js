import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import * as actions from '../actions';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

jest.mock('uuid', () => ({
  v4: () => '<id>',
}));

describe('users actions', () => {
  it('exports constants', () => {
    expect(actions.RECIEVE_AUTH).toBeDefined();
    expect(actions.SIGNIN).toBeDefined();
    expect(actions.SIGNUP).toBeDefined();
    expect(actions.LOGIN).toBeDefined();
    expect(actions.LOGOUT).toBeDefined();
  });

  it('recieves auth', () => {
    expect(actions.recieveAuth()).toMatchSnapshot();
  });

  it('signs in', () => {
    const id = 123;
    expect(actions.signin(id)).toMatchSnapshot();
  });

  it('signs up', () => {
    const username = 'user';
    expect(actions.signup(username)).toMatchSnapshot();
  });

  it('logins new user', () => {
    const user = {
      id: '123',
      username: 'user',
    };

    const store = mockStore({
      users: {
        byId: {},
      },
    });

    store.dispatch(actions.login(user.username));

    expect(store.getActions()).toEqual([
      { type: actions.LOGIN, username: user.username },
      actions.signup(user.username),
    ]);
  });

  it('logins existing user', () => {
    const user = {
      id: '123',
      username: 'user',
    };

    const store = mockStore({
      users: {
        byId: {
          [user.id]: user,
        },
      },
    });

    store.dispatch(actions.login(user.username));

    expect(store.getActions()).toEqual([
      { type: actions.LOGIN, username: user.username },
      actions.signin(user.id),
    ]);
  });

  it('logouts', () => {
    expect(actions.logout()).toMatchSnapshot();
  });
});
