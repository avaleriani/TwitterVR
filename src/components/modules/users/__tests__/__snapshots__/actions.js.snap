// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`users actions logouts 1`] = `
Object {
  "type": "@@users/LOGOUT",
}
`;

exports[`users actions recieves auth 1`] = `
Object {
  "type": "@@users/RECIEVE_AUTH",
}
`;

exports[`users actions signs in 1`] = `
Object {
  "payload": Object {
    "id": 123,
  },
  "type": "@@users/SIGNIN",
}
`;

exports[`users actions signs up 1`] = `
Object {
  "payload": Object {
    "id": "<id>",
    "username": "user",
  },
  "type": "@@users/SIGNUP",
}
`;
